RELEASING.md

```diff
@@ -0,0 +1,39 @@
+# Release Process
+
+## Pre Release
+
+1. Make sure all relevant changes for this release are included under `Unreleased` section in `CHANGELOG.md` and are in language that non-contributors to the project can understand.
+
+2. Run the pre-release script. It creates a branch `pre_release_<new-tag>` and updates `CHANGELOG.md` with the `<new-tag>`:
```

Comment:
My biggest concern here, is right now you can easily author a release tag thru GitHub UI. For that - there is also new GH console utility, which can be used to automate all GH UI abilities, including creating a new tag + adding release notes. It may be unnecessary to have a separate script for that, if we say that we use the GitHub release process, documented here:

https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/managing-releases-in-a-repository

That way even a person without console, e.g. Program Manager or Project Manager, can efficiently work on creating a new release quickly, without running any special tooling. I think it's a big advantage.

Reply:
The  github UI is supposed to be used to publish a release as mentioned in final section of this document  (https://github.com/open-telemetry/opentelemetry-cpp/blob/c0ea1fd587e1be0d99a97d639a1f82994515000c/RELEASING.md#release). I can make it explicit if this is not clear in document. The idea behind the first two steps are to manage the changelog.md. I am fine removing the second step ( Tagging ), and let the tag be created automatically by github UI if that's the ask ?

Reply:
We can probably skip the auto-tagging, and suggest that Release Manager uses their best judgement to tag the needed release sequentially, using semver 2.0, on that exact branch they wanna tag. We can keep the auto-tagging part if we want to cover tagging nightlies (but I am not sure we want this, as it'll create a ton of tags). One imporant moment where we may need to get the tag of the repo is to generate the `version.h` from the tag, not auto-increment the tag itself.

Reply:
The steps to be followed as per this PR is
```
1. CHANGELOG.md is managed manually, and as part of raising any PR, the dev needs to add the summary of change under "Unreleased" section.
2. On day of release, Release manager decides on a tag say ( v0.0.1-beta).
3. He creates a branch ( pre_release_v0.0.1-beta) - this is done by running bash script.
4. He modified the CHANGELOG.md ( replacing Unreleased section with "<tag>-<date>", and adding empty "Unreleased" section on top for next release.) - Again, this is done by bash script
5. He creates a PR for these change log changes, and waits for it to be merged to master.
6.  Once PR is merged, he creates the above decided tag on the new merged commit hash.
7. He then goes to the github release UI page, and select "Create New Release". Specifies the already created tag in step 6, and takes the curated contents as release notes, and creates the release.
```
Please let me know what change we are expecting in these steps ? We can remove step 6 and let github release UI create the tag automatically, but wonder what will we achieve with that :)

Reply:
> One imporant moment where we may need to get the tag of the repo is to generate the version.h from the tag, not auto-increment the tag itself.

I was thinking about this. Should we be having the release version in `version.h`. We do have `OPENTELEMETRY_VERSION` in `version.h` which seems to be used for dynamic loading of plugin libraries.

Reply:
I think we absolutely must include the runtime detail (as in numeric and semver string version of the library) in the SDK header. This will be needed for exporters that must populate SDK version in the attributes, for analytics and for crash analysis / stability reasons.

Reply:
There has to be two version concepts:
- version for API stability `VER_MAJOR.VER_MINOR`
- rest of semver 2.0 version for analytics and tracking; as there could be bugfix in SDK without a need to bump-up ver major and ver minor. 3rd tuple is the build increment, whereas the first two should be used to determine if the library has ABI-breaking changes.

Reply:
e.g. no need to change the namespace version if no ABI breaking changes, but there is still a need to regenerate the build id information in the header, for runtime use.

Reply:
Could we also append the latest commit hash to the version number and embed it into the build binary? This will help troubleshooting for developers.

Reply:
I think we can have both - short semver 2.0 in there with three tuples, but also some sort of `OsBuild` or `BUILD_ID`, which will be the full semver 2.0 --- with commit hash after dash. In many ways it's easier / nicer to work with just X.Y.Z, but agree for remote crash analysis - in prebuilt SDK shipped in products, we definitely need to capture and track the full build id including hash.

Reply:
QQ:  Does the tagged version ever make its way into a header file such that, e.g. we can use it in UserAgent strings when we write metrics?

Reply:
> Does the tagged version ever make its way into a header file such that, e.g. we can use it in UserAgent strings when we write metrics?

@jsuereth - Absolutely! My opinion we must have it and it could be used as such. In some future OpenTelemetry exporters, that my team would be working on, it is absolutely necessary to identify the version of the instrumentation library in some dedicated field. For both - user engagement analytics and crash analytics. Generated version info as both string and numeric constant, is a good starting point for enabling that. We also can reuse that for DLLs, like - there is a way to include a C header in a resource file, to be embedded into DLL on Windows. So when you right-click on a DLL - you see precisely what version it is. Usually just `X.Y.Z` but some projects include the full SemVer 2.0 with commit ID.

Reply:
Yes. I have a git issue to add version information into header file.